### PR TITLE
WIP - add incremental buffer updates

### PR DESCRIPTION
**Pull request by nosami**
Not ready for merging - still have some more work to do on this, i.e. every test is broken :)

### CODE REVIEW

File: omnisharp.el [hunk 1]
```diff
@@ -54,6 +55,12 @@ server backend."
   (omnisharp--init-imenu-support)
   (omnisharp--init-eldoc-support)
 
+  (add-hook 'before-change-functions
```
Comment (nosami) [inline omnisharp.el]: not sure if this is the right place to put this.

**Comment by nosami** [comment]
There are still some issues that I haven't worked out yet. 

Sometimes the buffer can get out of sync with the server. It is very easy to be out of sync if you make changes from emacs before starting the server for example. I've also seen it get out of sync at other times. This usually manifests itself as incorrect flycheck errors or wrong completions.

**Comment by nosami** [comment]
`revert-buffer` doesn't send change events. This could be fixed by having a server command to reload a buffer from disk.

**Comment by nosami** [comment]
revert-buffer now syncs the server from disk. It's dependent on https://github.com/OmniSharp/omnisharp-roslyn/pull/284 being merged in.

**Comment by nosami** [comment]
Rename fix dependant on https://github.com/OmniSharp/omnisharp-roslyn/pull/289

**Comment by razzmatazz** [comment]
@corngood – not sure if this is needed for roslyn rollout to `master` – seems like an awful lot of work to grasp this PR and fix it to a stable state and `omnisharp-emacs` w/roslyn would still work pretty good without it for the moment..

**Comment by nosami** [comment]
@corngood, @razzmatazz  From what I remember, this was just to allow the stdio interface to work with buffers over 4k in size. Also made editing _much_ faster when working with very large buffers

**Comment by razzmatazz** [comment]
@nosami do you think this is a big thing to make it work? for someone who did not read code for this PR before.. or do you still plan to work on it? as I understand then that this is a requirement for roslyn version of emacs so it would still work on large files?

**Comment by nosami** [comment]
@razzmatazz : I have no plans to work on it. I don't use omnisharp (or C#) any more.

You don't need this for roslyn if you carry on using omnisharp's HTTP interface.

I can't remember what state it was in either :( I think it was mostly working and just needed some more testing, but not sure at all now. Maybe @sp3ctum remembers?

**Comment by razzmatazz** [comment]
@nosami – we going to default on stdio now for roslyn, with http as an alternative i-face on special cases

do you know what is the underlying problem for 4k buffers? is it emacs, roslyn server or omnisharp-emacs code?

**Comment by nosami** [comment]
I seem to remember that there were 2 separate issues that made one big problem. Emacs can't store more than 4k in the pipe, and the Mono version of Console.ReadLineAsync was blocking at the time (maybe this has changed by now)

**Comment by razzmatazz** [comment]
Hmm.. I might not know the context well for this PR, but say, even with this incremental buffer -- wouldn't there still arise a situation where a big refactoring/code formatting/indentation change on a single large file STILL send more than 4K bytes to/from the server?

Wouldn't you still need to fix underlying issue on emacs and server-side?

Just theorising..

**Comment by nosami** [comment]
There is a check to see if the change > 4k. If it is, it writes to disk and the server reads from disk.

fwiw, I used the HTTP interface in omnisharp-roslyn with omnisharp-emacs master to develop omnisharp-roslyn itself. 

The main reason for wanting to use stdio was so that the server could push information to the editor without the editor requesting it (think background compilation, error messages, status updates etc). It also makes having multiple instances open a bit easier to manage (don't need to worry about finding free ports etc.)

**Comment by kimgronqvist** [comment]
I'm fairly sure the 4k issue is fixed already. It was reported here: 
https://debbugs.gnu.org/cgi/bugreport.cgi?bug=22344
https://github.com/flycheck/flycheck/issues/794

**Comment by razzmatazz** [comment]
@nosami @kimgronqvist 

Then I don't see it becoming showstopper and I have not encountered this problem myself (on 24.4/25.1+).

Due to time and resource constraints I will postpone fixup and merge of this PR until after `roslyn` version of omnisharp-emacs is released (4.0).

Unless someone complains this being a problem for them lately.

**Comment by sp3ctum** [comment]
I have a vague memory of there being some issue with how snippets are
expanded in Emacs. I think @nosami explained that whenever a character is
inserted or changed in the buffer, Emacs has a chance to send the changed
part of the buffer to Omnisharp via some Emacs hook.

But there was a special case with how snippets are handled: when a snippet
is expanded, the changes in the buffer can only be captured after all of
the snippet's fields are complete.

Remember that in company-mode it's possible to complete method calls so
that you automatically insert the entire signature, and then fill a
parameter, tab to the next parameter, etc.

So the problem for the user was that auto completion didn't work when
completing these snippets field by field. It's easy to see this because I'm
such an error state the buffer is not in sync with what the server knows of
the buffer.

Now, I'm not actually sure about this.. but it's what I gather the only
remaining problem was at the time. In any case it's worth taking a look at.

**Pull request closed without merging**